rosbag2_transport/src/rosbag2_transport/recorder.cpp

```diff
@@ -294,6 +302,56 @@ void RecorderImpl::record()
       writer_->split_bagfile();
     });
 
+  srv_start_discovery_ = node->create_service<rosbag2_interfaces::srv::StartDiscovery>(
+    "~/start_discovery",
+    [this](
+      const std::shared_ptr<rmw_request_id_t>/* request_header */,
+      const std::shared_ptr<rosbag2_interfaces::srv::StartDiscovery::Request>/* request */,
+      const std::shared_ptr<rosbag2_interfaces::srv::StartDiscovery::Response>/* response */)
+    {
+      start_discovery();
```

Comment:
Need to check `if (in_recording_)` before calling start discovery. Otherwise, it will be an undefined behavior since the writer may not be opened yet, but discovery can try to create a subscription with a callback calling `writer->write(message)`.